rts/motoko-rts/src/memory/ic.rs

```diff
@@ -89,12 +93,19 @@ impl Memory for IcMemory {
     }
 }
 
-/// Page allocation. Ensures that the memory up to, but excluding, the given pointer is allocated.
+/// Page allocation. Ensures that the memory up to, and including, the given pointer is allocated,
+/// with the slight exception of not allocating the extra page for 0x100_000_000. For all
+/// `ptr > 0x100_000_000` the call to `wasm32::memory_grow` will trap.
+/// Invariant: ptr must be a reasonable pointer (i.e. less than 0x200_000_000).
```

Comment:
```suggestion
/// Page allocation. Ensures that the memory up to, and including, the given pointer is allocated,
/// with the slight exception of not allocating the extra page for 0x1_0000_0000. For all
/// `ptr > 0x100_000_000` the call to `wasm32::memory_grow` will trap.
/// Invariant: ptr must be a reasonable pointer (i.e. less than 0x2_0000_0000).
```
Can you explain why this has to change? I'm a bit confused.

Reply:
I think this would be a good place to exclude the last page from being allocated, i.e. 0xFFFF_0000 should be the limit address for `HP`. Then we have a last-resort headroom for e.g. serialisation buffer (stable variables; caveat: do bignums allocate beyond that?)

Reply:
This would also render #3782's stack canary potentially unnecessary. Or it could be reduced to only be present in `--check` builds.

Reply:
Last page is free restriction is introduced now.